### PR TITLE
Charlie 167

### PR DESCRIPTION
**Pull request by zhalninde**

### CODE REVIEW

File: solana-common/src/main/kotlin/com/rarible/protocol/solana/common/repository/RecordsBalanceRepository.kt [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+        if (type.isNotEmpty()) {
+            val types = type.joinToString("|") {
+                when (it) {
+                    ActivityTypeDto.MINT -> "MintToRecord"
+                    ActivityTypeDto.BURN -> "BurnRecord"
+                    ActivityTypeDto.TRANSFER -> "TransferIncomeRecord"
+                }
+            }
+            criteria.and("_class").regex(".*($types)")
```
Comment (yakischik) [inline solana-common/src/main/kotlin/com/rarible/protocol/solana/common/repository/RecordsBalanceRepository.kt:47]: This query will work VERY slowly, should be IN of specific values.
Additional index required mint/_class/date/id

File: solana-common/src/main/kotlin/com/rarible/protocol/solana/common/repository/RecordsBalanceRepository.kt [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        } else {
+            return emptyFlow()
+        }
+
+        if (continuation != null) {
+            if (sort == ActivitySortDto.LATEST_FIRST) {
```
Comment (yakischik) [inline solana-common/src/main/kotlin/com/rarible/protocol/solana/common/repository/RecordsBalanceRepository.kt:53]: sort should be by date/_id, not only by _id

File: solana-common/src/main/kotlin/com/rarible/protocol/solana/common/repository/RecordsBalanceRepository.kt (hunk 1 above)
Comment (yakischik) [inline solana-common/src/main/kotlin/com/rarible/protocol/solana/common/repository/RecordsBalanceRepository.kt:47]: As far as i remember, criteria is immutable and this 'and' return new criteria

File: solana-api/src/main/kotlin/com/rarible/protocol/solana/nft/api/service/ActivityApiService.kt [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    ): ActivitiesDto {
+        val actualSize = size ?: 50
+
+        val activities = recordsBalanceRepository
+            .findByItem(type, tokenAddress, continuation, size, sort)
+            .mapNotNull { convert(it) }
+            .toList()
+
+        val outContinuation = activities
+            .takeIf { activities.size >= actualSize }
```
Comment (yakischik) [inline solana-api/src/main/kotlin/com/rarible/protocol/solana/nft/api/service/ActivityApiService.kt:30]: Continuation should be based on date/_id

File: solana-common/src/main/kotlin/com/rarible/protocol/solana/common/repository/RecordsBalanceRepository.kt (hunk 1 above)
Comment (zhalninde) [reply solana-common/src/main/kotlin/com/rarible/protocol/solana/common/repository/RecordsBalanceRepository.kt:47]: this dto has been moved and incorrect paths are stored in mongo, but now it seems possible to remove the regexps
index here:
https://github.com/rarible/solana-indexer/pull/9/files#diff-e9bfade7fb3e05db4097d7f7d6001a9f1e5b0368a98fce7e8aeb44655062b5fd

Comment (zhalninde) [reply solana-common/src/main/kotlin/com/rarible/protocol/solana/common/repository/RecordsBalanceRepository.kt:47]: criteris has a mutable state in criteriaChain member, this code works as expected

Comment (serejke) [reply solana-common/src/main/kotlin/com/rarible/protocol/solana/common/repository/RecordsBalanceRepository.kt:47]: please do not use regexps, list all needed classes

Comment (serejke) [reply solana-common/src/main/kotlin/com/rarible/protocol/solana/common/repository/RecordsBalanceRepository.kt:47]: This is not a problem, we will clean up the database. Use `in` query

File: solana-indexer/src/main/kotlin/com/rarible/protocol/solana/nft/listener/repository/BalanceLogRepository.kt [hunk 4]
```diff
@@ -60,8 +60,13 @@ class BalanceLogRepository(
             .sparse()
             .background()
 
+        private val BALANCE_CLASS: Index = Index()
+            .on("_class", Sort.Direction.ASC)
+            .background()
```
Comment (serejke) [inline solana-indexer/src/main/kotlin/com/rarible/protocol/solana/nft/listener/repository/BalanceLogRepository.kt:65]: Also add `on("_id", Sort.Direction.ASC)` here

File: solana-common/src/main/kotlin/com/rarible/protocol/solana/common/repository/RecordsBalanceRepository.kt (hunk 2 above)
Comment (zhalninde) [reply solana-common/src/main/kotlin/com/rarible/protocol/solana/common/repository/RecordsBalanceRepository.kt:53]: as far as I understood from the discussion with @serejke, _id is arranged in such a way that it provides sorting in chronological order

File: solana-common/src/main/kotlin/com/rarible/protocol/solana/common/repository/RecordsBalanceRepository.kt [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+@Component
+class RecordsBalanceRepository(
+    private val mongo: ReactiveMongoOperations,
+) {
+
+    suspend fun save(record: SolanaBalanceRecord): SolanaBalanceRecord =
+        mongo.save(record, COLLECTION).awaitSingle()
+
+    fun findAll() = mongo.findAll(SolanaBalanceRecord::class.java, COLLECTION).asFlow()
```
[the remark this answers is not in the document]
Comment (zhalninde) [reply solana-common/src/main/kotlin/com/rarible/protocol/solana/common/repository/RecordsBalanceRepository.kt:25]: +

File: solana-common/src/main/kotlin/com/rarible/protocol/solana/common/repository/RecordsBalanceRepository.kt [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+    fun findBy(query: Query): Flow<SolanaBalanceRecord> =
```
[the remark this answers is not in the document]
Comment (zhalninde) [reply solana-common/src/main/kotlin/com/rarible/protocol/solana/common/repository/RecordsBalanceRepository.kt:27]: +

**Pull request merged**